### PR TITLE
Trie implementation

### PR DESCRIPTION
**Pull request by ChinthanBD**

### CODE REVIEW

File: trie.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# https://www.codingninjas.com/studio/problems/implement-trie_631356?leftPanelTabValue=PROBLEM
+from sys import stdin, setrecursionlimit
+setrecursionlimit(10**7)
```
Comment (tejas-k3) [inline trie.py:3]: Why system's recursion limit is being changed?

File: trie.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class Node:
+    def __init__(self):
+        self.links = [None] * 26
+        self.flag = False
```
Comment (tejas-k3) [inline trie.py]: Flag is a generic term, what it means here?

File: trie.py (hunk 1 above)
Comment (ChinthanBD) [reply trie.py:3]: this was given implictly in the question link mentioned above

File: trie.py (hunk 2 above)
Comment (ChinthanBD) [reply trie.py]: Flag ending indicates whether the current char is the word ending or not

File: trie.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        node = self.root
+        for char in word:
+            pos = ord(char) - ord('a')
+            if not node.links[pos]:
+                return False
+            node = node.links[pos]
+
+        if node.flag == False:
+            return False
+        return True
```
Comment (tejas-k3) [inline trie.py]: just return node.flag, why overcomplicate?

File: trie.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        
+    def startWith(self, prefix):
+        node = self.root
+        for char in prefix:
+            pos = ord(char) - ord('a')
+            if not node.links[pos]:
+                return False
+            node = node.links[pos]
+        return True
```
Comment (tejas-k3) [inline trie.py:43]: this also can be simplified using `is not None` check

**Review by tejas-k3 — CHANGES_REQUESTED** [review]
Variable naming is too confusing for Trie

File: trie.py (hunk 2 above)
Comment (tejas-k3) [reply trie.py]: Then isn't `end_of_word` or `end` is better indicator?

File: trie.py (hunk 1 above)
Comment (tejas-k3) [reply trie.py:3]: Pls write comment for it that why it's required

File: trie.py (hunk 4 above)
Comment (ChinthanBD) [reply trie.py:43]: I feel the existing one is fine since we need to exit if link to any of the char is not there

File: trie.py (hunk 3 above)
Comment (ChinthanBD) [reply trie.py]: Done

File: trie.py (hunk 2 above)
Comment (ChinthanBD) [reply trie.py]: Done

File: trie.py (hunk 1 above)
Comment (ChinthanBD) [reply trie.py:3]: Done

Comment (tejas-k3) [reply trie.py:3]: This comment doesn't explain why such changes are present in code. Comment should be explaining the purpose of below code and where to use it, where to not use it
```
from sys import stdin, setrecursionlimit
setrecursionlimit(10**7)
```

**Review by tejas-k3 — CHANGES_REQUESTED** [review]
Add 1 more method which should print list of all words which start with given prefix and try to reuse `startWith` method with modifications IF required

File: trie.py (hunk 1 above)
Comment (ChinthanBD) [reply trie.py:3]: The recursion limit is set to 10**7  to prevent the program from running out of memory or encountering a stack overflow error.

Comment (tejas-k3) [reply trie.py:3]: Running out of memory is mostly a good idea, you can't have your program taking more than a limited memory else it'll cause system failure. And same follows for stack overflow error unless you're very clear about pros and cons of it.

Comment (tejas-k3) [reply trie.py:3]: By default, python limits max stack for recursion to 1000 recursive calls. You can confirm it by
```
import sys
print(sys.getrecursionlimit())
```
For this particular code, as all the operations are using **Iterative approach** it makes no difference but IF you were to use recursive calls for say insert or search (Example -> Search in reverse order) then with python's current limit of 1000 being stack limit.. Any word having char length more than 1000 would end up causing 1000+ recursive calls leading to program failure. If you implement any method using **Recursive approach** where valid computation can go beyond 1000 recursive calls (word having 1000+ chars for this example), you need to acc resize the system stack limit. This is all constraint based. This is one of many reasons why VALID DFS based questions dont contain testcases or constraints which can lead up to 1000 >= recursive calls

**Pull request merged**